### PR TITLE
feat: add semantic version utilities for e2e tests

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: e2e/testsuite/semver_test.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+		{"non semantic version", "1.minor.1", true},
+		{"supported semantic version", "v2.5.0", true},
+		{"supported semantic version", "v3.4.0", true},
+		{"supported semantic version", "v4.2.0", true},
+		{"supported semantic version", "v5.1.0", true},
+		{"supported semantic version", "v6.0.0", true},
+		{"supported semantic version", "v6.1.0", true},
+		{"supported semantic version", "v7.1.0", true},
+		{"supported semantic version", "v22.5.1", true},
+		{"supported semantic version with v", "2.5.0", true},
```
Comment (crodriguezvega) [inline e2e/testsuite/semver_test.go:40]:
```suggestion
		{"supported semantic version without v", "2.5.0", true},
```

File: e2e/testsuite/semver.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+// NewVersion constructs a new version given the major and minor version.
+func NewVersion(major, minor uint64) Version {
+	return Version{
+		Major: major,
+		Minor: minor,
+	}
+}
+
+func ParseChainVersion(version string) (Version, bool) {
```
Comment (crodriguezvega) [inline e2e/testsuite/semver.go:22]: Is this function is only used in `IsSupported`, can it be unexported? If it cannot, can we add a godoc then?

File: e2e/testsuite/semver.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	MajorVersion  uint64
+	MinorVersions []Version
+}
+
+// IsSupported returns whether the version contains the feature.
+// This is true if the version is greater than or equal to the major version it was released in
+// or is greater than or equal to the list of minor releases it was included in.
+func (fr FeatureReleases) IsSupported(versionStr string) bool {
+	// assume any non semantic version formatted version support the feature
+	// this will be useful during development of the e2e test with the new feature
```
Comment (crodriguezvega) [inline e2e/testsuite/semver.go:58]: So this is to support images with names of working branches, right?

File: e2e/testsuite/semver.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	minor, err := strconv.ParseUint(versions[1], 10, 64)
+	if err != nil {
+		return Version{}, false
+	}
+
+	return Version{
+		Major: major,
+		Minor: minor,
+	}, true
```
Comment (chatton) [inline e2e/testsuite/semver.go:44]: The majority of the parsing logic here seems to be duplicating  `semver.Major` and `semver.Minor`

File: e2e/testsuite/semver.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+	if version.Major >= fr.MajorVersion {
+		return true
+	}
+
+	for _, mv := range fr.MinorVersions {
+		if mv.Major == version.Major {
+			return version.Minor >= mv.Minor
+		}
+	}
```
Comment (chatton) [inline e2e/testsuite/semver.go:72]: as the input to this funciton is the version string, we can use `semver.Compare` to handle the version comparisons.

**Review by chatton — COMMENTED** [review]
The test cases look great ❤️ 

I think a lot of the custom logic that we have here can be replaced with the `semver` package. Especially since our inputs are strings.

File: e2e/testsuite/semver.go (hunk 4 above)
Comment (colin-axner) [reply e2e/testsuite/semver.go:44]: ah! I was looking for that, wasn't sure it existed, looks to be importable under "semver"?

**Comment by chatton** [comment]
After the refactor, the custom types to represent versions were removed in favour of strings which are natively supported by the `semver` package. A few small functions were added to aid readability. All test cases remain the same.

**Review by colin-axner — COMMENTED** [review]
ACK, thanks @chatton! Feel free to merge for me whenever you feel ready

File: e2e/testsuite/semver.go (hunk 3 above)
Comment (crodriguezvega) [reply e2e/testsuite/semver.go:20]: All right... maybe this is an edge case, but if we have a PR for a branch that does not support the feature (for example, a backport to `release/v2.4.x`) then the test will fail, right? Because this will wrongly assume that the PR supports the feature... Although maybe this is not a problem if the metadata tests are only run as part of the compatibility check.

**Review by charleenfei — COMMENTED** [review]
nice work @colin-axner @chatton !

**Review by colin-axner — COMMENTED** [review]
Approve! (can't click approve button on github, feel free to do it on my behalf)

**Review by chatton — APPROVED** [review]
(on behalf of @colin-axner  ❤️ )
